### DISCUSSION TITLE
TypeScript - Pass mutation to function

### DISCUSSION THREAD

**Discussion by BrendanC23** [trpc/trpc#5586] [Q&A]
My router has several different routes with a similar signature. For example: `(data: string) => string`.

```ts
const mutation1 = trpc.router.insertItemA.useMutation();
mutation`.mutate({ data: "name" }, { onSuccess: (res: string) => {});

const mutation2 = trpc.router.insertItemA.useMutation();
mutation2.mutate({ data: "name" }, { onSuccess: (res: string) => {});
```

I want to do something like the following, where `fn` would accept a parameter of either `mutation1` or `mutation2`.

```ts
function fn(mutation: Mutation<(data: string) => string>) {}
```

I see that TRPC has a `MutationLike` [type](https://github.com/trpc/trpc/blob/main/packages/react-query/src/shared/polymorphism/mutationLike.ts), but I can't find any examples about how to actually use it.

**Comment by CHC383** [comment]
That's the interface in v10, so you can use it in these two ways:
- `trpc.router.insertItemA` -> `MutationLike`
-  `trpc.router.insertItemA.useMutation()` -> `MutationLike["useMutation"]`

It is similar for `QueryLike` or `QueryLike["useQuery"]`, or you could even have `queryInput` typed as `Parameters<QueryLike["useQuery"]>[0]`, `queryOpts` typed as `Parameters<QueryLike["useQuery"]>[1]`

However v11 has the interface changed, I am searching around for that without any findings yet, and found this question
